enable marking as spam

jackiehluo opened the pull request:
adds the ability to show/hide collections in the collection editor!

collected view:
<img width="242" alt="Screen Shot 2022-11-11 at 2 02 58 PM" src="https://user-images.githubusercontent.com/8452682/201415885-3c173d9e-0917-4f9c-817d-f8427de4a788.png">

hidden view:
<img width="240" alt="Screen Shot 2022-11-11 at 2 23 52 PM" src="https://user-images.githubusercontent.com/8452682/201415896-426f4e98-bd13-4b50-9cb2-41272d30f811.png">

selector open state:
<img width="238" alt="Screen Shot 2022-11-11 at 2 03 15 PM" src="https://user-images.githubusercontent.com/8452682/201415906-2febc185-8fa9-4435-bb44-11e1e9850840.png">

tbezman commented on src/components/ManageGallery/OrganizeCollection/Sidebar/SidebarTokens.tsx:
```diff
@@ -45,6 +50,26 @@ export const SidebarTokens = ({
+  const setSpamPreference = useSetSpamPreference();
+  const setSpamPreferenceForCollection = useCallback(
+    (address: string, isSpam: boolean) => {
+      const groups = groupCollectionsByAddress({ tokens, editModeTokens });
+      const matchingCollection = groups.find((group) => group.address === address);
```
correct me if i'm wrong but I think there's some indirection here. 

we have an address, and a list of tokens. we then group the tokens by address into a list. then we search for that group to get the tokens.

I think we can just filter the list of tokens by their contract address.

jackiehluo replied:
oh nope haha you are 100% right! thanks for catching that, will update

tbezman commented on src/components/ManageGallery/OrganizeCollection/Sidebar/SidebarViewSelector.tsx:
```diff
@@ -0,0 +1,55 @@
+export function SidebarViewSelector({ selectedView, setSelectedView }: SidebarViewSelectorProps) {
+  const [isDropdownOpen, setIsDropdownOpen] = useState(false);
+
+  const onSelectView = (selectedView: SidebarView) => {
+    setSelectedView(selectedView);
+    setIsDropdownOpen(false);
+  };
```
nit: usecallback

tbezman commented on src/components/core/Dropdown/Dropdown.tsx:
```diff
@@ -10,9 +10,10 @@ type Props = {
   onClose: () => void;
   children?: ReactNode;
   position: 'right' | 'left';
+  style?: React.CSSProperties;
```
maybe instead of passing in a custom `style`, we update `position` to have another `full-width` option to compete with `left` and `right`.

I imagine this won't be the last time we want a full width version of the dropdown

Robinnnnn commented on src/components/ManageGallery/OrganizeCollection/Sidebar/SidebarList.tsx:
```diff
@@ -203,3 +282,17 @@ const Selection = styled.div`
+  right: 0;
+  top: 30px;
+  z-index: 1;
+`;
```
could we make the tooltip a bit wider? it's 130px in figma

<img width="325" alt="image" src="https://user-images.githubusercontent.com/12162433/201436743-d78e7992-2def-477f-a14c-6e4331466244.png">

wondering if we could position it this way as well:

<img width="130" alt="image" src="https://user-images.githubusercontent.com/12162433/201436786-0ffefad6-d92f-4fce-a9dd-fc71a6d69d8e.png">

jackiehluo replied:
ooh yes!

and re: positioning, i think `<Tooltip />` would need to be reworked to use a react portal so that it can render outside of the sidebar! i was worried it might affect other places it's used in unexpected ways, but i can def add that—thoughts? i think it does make sense to do that in general, probably also for `<Dropdown />`

an easier/more reliable way to handle it versus rolling it from scratch would be using a library (i.e., https://floating-ui.com/docs/floatingportal) that handles positioning, focus, etc., out of the box

Robinnnnn replied:
ahh wow forgot about the sidebar, and great point about using a portal - that sounds great!

i don't think this is blocking, so i'll create a separate ticket with the floating portal suggestion as a fast-follow :) no need to handle in this PR

Robinnnnn commented on src/hooks/api/tokens/useSetSpamPreference.ts:
```diff
@@ -0,0 +1,30 @@
+    `
+  );
+
+  return useCallback(
+    async (input: SetSpamPreferenceInput) => {
+      await setSpamPreference({ variables: { input } });
+
+      // TODO: handle error cases
+    },
```
we'll want to make sure this is updated optimistically so that the collection is immediately moved to the hidden folder. @tbezman can you point to any good precedents in the codebase?

jackiehluo replied:
haha oh yeah i had a few ideas of how to do it but all of them felt janky—but just now reading the relay docs more and seeing `optimisticResponse`

Robinnnnn reviewed (commented):
for the empty view on the hidden tab, could we just toss the copy? 

<img width="252" alt="Screen Shot 2022-11-11 at 5 06 20 PM" src="https://user-images.githubusercontent.com/12162433/201437446-c384ac85-e305-4b94-951d-2b19d3a0500c.png">

jackiehluo commented on src/components/ManageGallery/OrganizeCollection/Sidebar/SidebarViewSelector.tsx:
```diff
@@ -0,0 +1,55 @@
+    <Container>
+      <Selector onClick={() => setIsDropdownOpen(true)}>
+        <BaseM>{selectedView}</BaseM>
+        <Arrows style={{ marginRight: 4 }} />
```
oooh okay yes missed this whole subdirectory! okay sweet

jackiehluo commented:
okay i addressed most of the comments! only parts unresolved are tooltip positioning and optimistic updates (i added an `optimisticResponse`, but as far as i can tell, it's silently failing? @tbezman is there's anything obvious i missed?)

<img width="242" alt="Screen Shot 2022-11-12 at 12 17 57 PM" src="https://user-images.githubusercontent.com/8452682/201486474-a96b4d37-b9a3-481f-ac81-5a3bb9cd10e1.png">

jackiehluo commented:
fixed the optimistic updating with @tbezman's help! thinking more about the tooltip, i think that's probably best suited for its own project and out of scope for this task?

Robinnnnn reviewed (approved):
absolutely crushed it 💪 💪 💪 💪
